lib/puppet/provider/bootstrap_rpm/bootstrap_rpm.rb

```diff
@@ -0,0 +1,139 @@
+require 'fileutils'
+
+Puppet::Type.type(:bootstrap_rpm).provide(:bootstrap_rpm) do
+
+  commands :rpmbuild => 'rpmbuild'
+
+  def create
+    base_dir
+    copy_sources
+    write_specfile
+    build_rpm
+    copy_rpm
+    link_rpm
+  ensure
+    FileUtils.remove_dir(base_dir) if File.directory?(base_dir)
+  end
+
+  def delete
+  end
+
+  def exists?
+    File.exist?(resource[:symlink]) if resource[:symlink]
+  end
+
+	private
+
+  def base_dir
+    @base_dir ||= Dir.mktmpdir
+  end
+
+  def sources_dir
+    mkdir('SOURCES')
+  end
+
+  def build_dir
+    mkdir('BUILD')
+  end
+
+  def spec_dir
+    mkdir('SPECS')
+  end
+
+  def rpm_dir
+    mkdir('RPMS')
+  end
+
+  def copy_sources
+    FileUtils.copy(resource[:script], "#{sources_dir}/#{File.basename(resource[:script])}")
+  end
+
+  def write_specfile
+    File.open("#{spec_dir}/#{resource[:name]}.spec", 'w') do |file|
+      file.write(spec)
+    end
+  end
+
+  def release
+    if !latest_rpm.nil? && File.exist?(latest_rpm)
+      rel = `rpm -qp #{latest_rpm} --queryformat=%{release}`
+      rel.to_i + 1
+    else
+      1
+    end
+  end
+
+  def build_rpm
+    command = ['rpmbuild', '-ba']
```

Comment:
You should also use `rpmbuild()` here to call the right command. Backticks don't use the proper error handling. You also get more debugging.

Reply:
I tried that, whatever is underlying that functionality does not play nicely with rpmbuild. I had similar issues with shelling out in python to call rpmbuild in Obal. The macros don't get passed to rpmbuild correctly:

```
  Info: Foreman_proxy_content::Rhsm_reconfigure_script[/var/www/html/pub/katello-rhsm-consumer]: Scheduling refresh of Bootstrap_rpm[katello-ca-consumer-centos7-64.example.com]
  Error: Execution of '/usr/bin/rpmbuild -ba /tmp/d20200714-1701-os51rg/SPECS/katello-ca-consumer-centos7-64.example.com.spec --define "_topdir /tmp/d20200714-1701-os51rg" --define "name katello-ca-consumer-centos7-64.example.com" --define "release 1"' returned 1: error: Macro % has illegal name (%define)
  error: Macro % has illegal name (%define)
  error: Macro % has illegal name (%define)
  error: Macro % has illegal name (%define)
  error: Macro % has illegal name (%define)
  error: Macro % has illegal name (%define)
  error: Macro % has illegal name (%define)
  error: Too many levels of recursion in macro expansion. It is likely caused by recursive macro declaration.
```

Reply:
I think that's because you double quote it. I'll take a stab at this.

Reply:
This works for me:
```ruby
  def build_rpm
    rpmbuild('-ba', File.join(spec_dir, "#{resource[:name]}.spec"),
             '--define', "_topdir #{base_dir}",
             '--define', "name #{resource[:name]}",
             '--define', "release #{release}")
  end 
```